### PR TITLE
feat: forbid to set matchLabelKeys when labelSelector isn’t set in topologySpreadConstraints

### PR DESCRIPTION
**Pull request by denkensk**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
When users set `lableSelector` in topologySpreadConstraints as "null", it represents `labels.Nothing()`, so the key in matchLabelKeys needs to be ignored. 
```yaml
     topologySpreadConstraints:
        - maxSkew: 1
          topologyKey: kubernetes.io/hostname
          whenUnsatisfiable: DoNotSchedule
          matchLabelKeys:
            - app
            - pod-template-hash
```

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

It's ok to change the behavior because the feature is graduating to beta in 1.27

#### Does this PR introduce a user-facing change?

```release-note
forbid to set matchLabelKeys when labelSelector isn’t set in topologySpreadConstraints
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by sanposhiho** [comment]
@denkensk it's better to have a release note since this change may break someone's PodTopologySpread with MatchLabelKeys.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go [hunk 1]
```diff
@@ -1299,6 +1299,45 @@ func TestPreFilterState(t *testing.T) {
 			},
 			enableMatchLabelKeys: true,
 		},
+		{
+			name: "key in matchLabelKeys is ignored when LabelSelector is nil",
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:1303]: Could you add the test case "no match with any Pods when LabelSelector is nil"? (not directly related to MatchLabelKeys though)

**Comment by sanposhiho** [comment]
/triage accepted
/priority important-soon

(Probably, we want to merge this before code-freeze since the matchLabelKeys in TopologySpread is gonna be beta and get enabled by default in v1.27)

File: pkg/scheduler/framework/plugins/podtopologyspread/common.go [hunk 2]
```diff
@@ -139,6 +139,8 @@ func mergeLabelSetWithSelector(matchLabels labels.Set, s labels.Selector) labels
 		for _, r := range requirements {
 			mergedSelector = mergedSelector.Add(r)
 		}
+	} else {
+		return s
```
[the remark this answers is not in the document]
Comment (denkensk) [reply pkg/scheduler/framework/plugins/podtopologyspread/common.go:143]: Updated.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go (hunk 1 above)
Comment (denkensk) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go]: Done

**Comment by alculquicondor** [comment]
Add in notes to reviewer:

It's ok to change the behavior because the feature is graduating to beta in 1.27

**Comment by denkensk** [comment]
> Add in notes to reviewer:
> 
> 
> 
> It's ok to change the behavior because the feature is graduating to beta in 1.27

Added

**Comment by alculquicondor** [comment]
@liggitt we need your input.

The question is whether they should have the same behavior:

```yaml
     topologySpreadConstraints:
        - maxSkew: 1
          topologyKey: kubernetes.io/hostname
          whenUnsatisfiable: DoNotSchedule
          matchLabelKeys:
            - app
```

```yaml
     topologySpreadConstraints:
        - maxSkew: 1
          topologyKey: kubernetes.io/hostname
          whenUnsatisfiable: DoNotSchedule
          labelSelector: {}
          matchLabelKeys:
            - app
```


These two mean different things:

Match none:
```yaml
     topologySpreadConstraints:
        - maxSkew: 1
          topologyKey: kubernetes.io/hostname
          whenUnsatisfiable: DoNotSchedule
```

Match all:
```yaml
     topologySpreadConstraints:
        - maxSkew: 1
          topologyKey: kubernetes.io/hostname
          whenUnsatisfiable: DoNotSchedule
          labelSelector: {}
```

**Comment by alculquicondor** [comment]
Right, if we want to respect that null labelSelector means match-none in all cases, forbidding a nil selector makes sense, otherwise pods are unschedulable.

Note that labelSelector means match-none for Deployment, Job, etc. So we are better of keeping that behavior.

**Comment by alculquicondor** [comment]
> (just to be clear, the thing we'd be forbidding is a non-zero matchLabels value when selector is nil, right?)

yes :)

**Comment by sanposhiho** [comment]
So, the conclusion here is
- we'll forbid nil selector with non-nil matchLabelKeys
- (And then we'll (probably) emit a warning in nil selector itself in https://github.com/kubernetes/kubernetes/issues/115873.)

Sounds good. I'll follow this direction in https://github.com/kubernetes/kubernetes/pull/116065 as well if no further concern or other better ideas.

**Comment by denkensk** [comment]
> Right, if we want to respect that null labelSelector means match-none in all cases, forbidding a nil selector makes sense, otherwise pods are unschedulable.

> Note that labelSelector means match-none for Deployment, Job, etc. So we are better of keeping that behavior.

Yes. It's better to keep the same behavior.

> we'd be forbidding is a non-zero matchLabels value when selector is nil

I will update this pr today.

**Comment by denkensk** [comment]
> @denkensk Could you have time to work on changing [`validateMatchLabelKeys`](https://github.com/sanposhiho/kubernetes/blob/b740a343029ab50751015b91bca81511bd94a0ff/pkg/apis/core/validation/validation.go#L6962) to forbid nil selector with non-nil matchLabelKeys in this PR? (If you don't have the bandwidth to work on it, just let me know, I'll implement it in #116065 or another PR. 🙂) (PodAffinity uses the same function for validation of matchLabelKeys.)

Yes. I can finish it today. Just woke up because of the time zone  ^_^

File: pkg/apis/core/validation/validation.go [hunk 3]
```diff
@@ -6964,17 +6964,20 @@ func validateMatchLabelKeys(fldPath *field.Path, matchLabelKeys []string, labelS
 		return nil
 	}
 
+	allErrs := field.ErrorList{}
+	if labelSelector == nil {
+		allErrs = append(allErrs, field.Forbidden(fldPath, "must not be specified when labelSelector is not set"))
+		return allErrs
+	}
```
Comment (sanposhiho) [inline pkg/apis/core/validation/validation.go:6971]: nits:

```suggestion
	if labelSelector == nil {
		return field.ErrorList{field.Forbidden(fldPath, "must not be specified when labelSelector is not set")}
	}
```

Comment (denkensk) [reply pkg/apis/core/validation/validation.go:6971]: 🤔 I prefer to keep it so that it is consistent with the style of the code in that file and also has better readability.

**Comment by denkensk** [comment]
@liggitt  please take a look. Thanks.

File: pkg/apis/core/validation/validation.go [hunk 4]
```diff
@@ -6964,17 +6964,20 @@ func validateMatchLabelKeys(fldPath *field.Path, matchLabelKeys []string, labelS
 		return nil
 	}
 
+	allErrs := field.ErrorList{}
+	if labelSelector == nil {
+		allErrs = append(allErrs, field.Forbidden(fldPath, "must not be specified when labelSelector is not set"))
+		return allErrs
```
Comment (alculquicondor) [inline pkg/apis/core/validation/validation.go:6970]: do not return, continue validating

File: pkg/scheduler/framework/plugins/podtopologyspread/common.go [hunk 5]
```diff
@@ -135,10 +135,14 @@ func (pl *PodTopologySpread) filterTopologySpreadConstraints(constraints []v1.To
 
 func mergeLabelSetWithSelector(matchLabels labels.Set, s labels.Selector) labels.Selector {
 	mergedSelector := labels.SelectorFromSet(matchLabels)
-	if requirements, ok := s.Requirements(); ok {
-		for _, r := range requirements {
-			mergedSelector = mergedSelector.Add(r)
-		}
+
+	requirements, ok := s.Requirements()
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/podtopologyspread/common.go:139]: I guess this change doesn't matter anymore?

Comment (denkensk) [reply pkg/scheduler/framework/plugins/podtopologyspread/common.go:139]: Actually Yes. But the change is still needed to ensure that the logic is correct

File: pkg/apis/core/validation/validation_test.go [hunk 6]
```diff
@@ -22091,6 +22091,7 @@ func TestValidateTopologySpreadConstraints(t *testing.T) {
 				{
 					MaxSkew:           1,
 					TopologyKey:       "k8s.io/zone",
+					LabelSelector:     &metav1.LabelSelector{MatchLabels: map[string]string{"foo": "foo"}},
```
[the remark this answers is not in the document]
Comment (denkensk) [reply pkg/apis/core/validation/validation_test.go]: Done

File: pkg/apis/core/validation/validation_test.go [hunk 7]
```diff
@@ -22118,6 +22119,18 @@ func TestValidateTopologySpreadConstraints(t *testing.T) {
...
+			name: "key in MatchLabelKeys is forbidden to be specified when labelSelector is not set",
+			constraints: []core.TopologySpreadConstraint{
+				{
+					MaxSkew:           1,
+					TopologyKey:       "k8s.io/zone",
+					WhenUnsatisfiable: core.DoNotSchedule,
+					MatchLabelKeys:    []string{"/simple"},
+				},
+			},
+			wantFieldErrors: []*field.Error{field.Forbidden(fieldPathMatchLabelKeys, "must not be specified when labelSelector is not set")},
```
[the remark this answers is not in the document]
Comment (denkensk) [reply pkg/apis/core/validation/validation_test.go]: Done

File: pkg/scheduler/framework/plugins/podtopologyspread/common.go (hunk 5 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/podtopologyspread/common.go:139]: right, so this would ensure that empty label selector means match-none.

/lgtm
/approve

**Comment by alculquicondor** [comment]
release note is out-of-date

/hold

**Comment by alculquicondor** [comment]
/lgtm cancel

Please also update the API field comment to state the validation rule

File: pkg/apis/core/types.go [hunk 8]
```diff
@@ -5972,8 +5972,12 @@ type TopologySpreadConstraint struct {
 	// spreading will be calculated. The keys are used to lookup values from the
 	// incoming pod labels, those key-value labels are ANDed with labelSelector
 	// to select the group of existing pods over which spreading will be calculated
-	// for the incoming pod. Keys that don't exist in the incoming pod labels will
+	// for the incoming pod. The same key is forbidden to exist in both MatchLabelKeys and LabelSelector.
+	// MatchLabelKeys is forbidden to set when LabelSelector isn't set.
```
Comment (alculquicondor) [inline pkg/apis/core/types.go:5976]:
```suggestion
	// MatchLabelKeys cannot be set when LabelSelector isn't set.
```

Comment (denkensk) [reply pkg/apis/core/types.go]: Done

**Comment by alculquicondor** [comment]
This is a bug fix, so it qualifies to be merged before test freeze.

**Comment by liggitt** [comment]
/approve
API doc and validation change lgtm (this field was alpha in 1.26, tightening validation before releasing as beta is in bounds)

@alculquicondor has lgtm and can cancel hold

**Comment by alculquicondor** [comment]
FYI: The website needs updates. The example doesn't have a node-selector https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/#spread-constraint-definition

/hold cancel

**Comment by denkensk** [comment]
> FYI: The website needs updates. The example doesn't have a node-selector https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/#spread-constraint-definition
> 
> /hold cancel

Update it in https://github.com/kubernetes/website/pull/40008
